api/src/wfl/module/covid.clj

```diff
@@ -58,16 +31,27 @@
       (throw (ex-info (str "No such record") {:id id :table table})))
     record))
 
-;; interfaces
-;; queue operations
-(defmulti peek-queue!
-  "Peek the first object from the `queue`, if one exists."
-  (fn [queue] (:type queue)))
-
-(defmulti pop-queue!
-  "Pop the first object from the `queue`. Throws if none exists."
-  (fn [queue] (:type queue)))
+(defn peek-queue!
+  "Get first unconsumed item in DETAILS table."
+  [{:keys [details] :as _queue}]
+  (let [query "SELECT * FROM %s WHERE NOT consumed ORDER BY id ASC LIMIT 1"]
+    (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+      (->> (format query details)
+           (jdbc/query tx)
+           (first)))))
+
+(defn pop-queue!
+  "Consume first unconsumed item in DETAILS table, or throw if empty."
+  [{:keys [details] :as queue}]
+  (if-let [{:keys [id] :as _item} (peek-queue! queue)]
+    (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+      (let [now (OffsetDateTime/now)]
+        (jdbc/update! tx details {:consumed now
+                                  :updated  now}
+                      ["id = ?" id])))
+    (throw (ex-info "No items in queue" {:queue queue}))))
```

Comment:
SELECT and UPDATE are in different transactions,
which is _probably_ OK here.

Reply:
We could change the queue methods to accept transactions, but I think that's an okay change to save for later.